### PR TITLE
Sunghcho/denormals

### PR DESCRIPTION
**Pull request by hanbitmyths**
**Description**: Add new session option and global threading option to set denormal as zero.

**Motivation and Context**
When a model contains denormal floating number in weight or input, computation cost is high and sometimes it's more than 30 times slower than without denormals. This PR introduces new session option and global threading option to be able to set denormal as zero.
One note is that flush-to-zero and Denormal-as-zero can be done at thread level. It means main thread and OpenMP threads can't be fully controlled by session option. only the first session option is effective to main thread and OpenMP threads.

### CODE REVIEW

**Review by tracysh — CHANGES_REQUESTED** [review]
We shouldn't need to pull in cpuinfo just to get the DAZ flag: we can add something to MLAS that exposes a bool for you to check instead.

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 1]
```diff
@@ -1088,7 +1088,13 @@ struct OrtApi {
   ORT_API2_STATUS(CreateEnvWithCustomLoggerAndGlobalThreadPools, OrtLoggingFunction logging_function, _In_opt_ void* logger_param, OrtLoggingLevel logging_level,
                   _In_ const char* logid, _In_ const struct OrtThreadingOptions* tp_options, _Outptr_ OrtEnv** out);
 
- #ifdef USE_CUDA
+  /**
+   * Use this API to configure the global thread pool options to be used in the call to CreateEnvWithGlobalThreadPools.
+   * Set denormal as zero.
+   */
+  ORT_API2_STATUS(SetGlobalDenormalAsZero, _Inout_ OrtThreadingOptions* tp_options);
```
Comment (yuslepukhin) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:1095]: >ORT_API2_STATUS(SetGlobalDenormalAsZ [](start = 2, length = 36)

New APIs can only be appended, never inserted to preserve ABI compatibility with C# and potentially others

Comment (hanbitmyths) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: Mistake on merging

**Comment by hanbitmyths** [comment]
> We shouldn't need to pull in cpuinfo just to get the DAZ flag: we can add something to MLAS that exposes a bool for you to check instead. We already do CPU detection from mlas\lib\platform.cpp, we can extend that.

@tracysh, I removed cpuinfo submodule and modified to detect only sse3 support like Tensorflow.

File: onnxruntime/core/common/denormal.cc [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  return false;
+}
+
+#ifdef _OPENMP
+class DenormalAsZeroInitializer {
+ public:
+  explicit DenormalAsZeroInitializer(bool on) : on_(on) {}
+
+  DenormalAsZeroInitializer(const DenormalAsZeroInitializer& init) : on_(init.on_) {
+    SetDenormalAsZero(on_);
```
Comment (pranavsharma) [inline onnxruntime/core/common/denormal.cc:52]: A bit confused. Why does the copy ctor call SetDenormalAsZero, but not the ctor?

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 3]
```diff
@@ -1094,6 +1094,12 @@ struct OrtApi {
   ORT_API2_STATUS(OrtSessionOptionsAppendExecutionProvider_CUDA,
                   _In_ OrtSessionOptions* options, _In_ OrtCUDAProviderOptions* cuda_options);
 #endif  // USE_CUDA
+
+  /**
+   * Use this API to configure the global thread pool options to be used in the call to CreateEnvWithGlobalThreadPools.
+   * Set denormal as zero.
```
Comment (pranavsharma) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:1100]: Would be good to document: 1. when would someone want to call this API, what alternatives exist and what's the impact of using this (like loss of accuracy)? 2. this applies to both inter and intra op threadpools.

File: onnxruntime/core/session/inference_session.cc [hunk 4]
```diff
@@ -202,6 +203,20 @@ void InferenceSession::ConstructorCommon(const SessionOptions& session_options,
...
+
+  // The only first session option for flush-to-zero and denormal-as-zero is effective to main thread and OpenMP threads.
+  {
+    static std::once_flag once;
+
+    std::call_once(once, [&] {
+#ifdef _OPENMP
+      InitializeWithDenormalAsZero(set_denormal_as_zero);
+#endif
+      SetDenormalAsZero(set_denormal_as_zero);
```
Comment (pranavsharma) [inline onnxruntime/core/session/inference_session.cc:216]: Would be good to log a line at INFO level saying that this setting is being turned on.

**Comment by pranavsharma** [comment]
Is the flatbuffers related change intended?

**Comment by hanbitmyths** [comment]
> Is the flatbuffers related change intended?

It's from merge, not from my change.

File: onnxruntime/core/common/denormal.cc (hunk 2 above)
Comment (hanbitmyths) [reply onnxruntime/core/common/denormal.cc:52]: It's a trick solution to run initialization code at all open mp threads. I'll write down comments here.

File: onnxruntime/core/common/denormal.cc [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+ private:
+  bool on_;
+};
+
+void InitializeWithDenormalAsZero(bool on) {
+  DenormalAsZeroInitializer init(on);
+#pragma omp parallel for firstprivate(init)
```
Comment (tracysh) [inline onnxruntime/core/common/denormal.cc:61]: From what I've seen with MSVC + openmp, if the parent thread that spawns off a parallel section does anything to change the number of threads used, then openmp recreates the task set. And the pool is per-thread, but I guess I need to see the comments describing how this works above.

File: onnxruntime/test/perftest/command_args_parser.cc [hunk 6]
```diff
@@ -46,16 +46,17 @@ namespace perftest {
       "\t-y [inter_op_num_threads]: Sets the number of threads used to parallelize the execution of the graph (across nodes), A value of 0 means ORT will pick a default. Must >=0.\n"
       "\t-P: Use parallel executor instead of sequential executor.\n"
       "\t-o [optimization level]: Default is 1. Valid values are 0 (disable), 1 (basic), 2 (extended), 99 (all).\n"
-      "\t\tPlease see onnxruntime_c_api.h (enum GraphOptimizationLevel) for the full list of all optimization levels. \n"
+      "\t\tPlease see onnxruntime_c_api.h (enum GraphOptimizationLevel) for the full list of all optimization levels.\n"
       "\t-u [optimized_model_path]: Specify the optimized model path for saving.\n"
       "\t-d [cudnn_conv_algorithm]: Specify CUDNN convolution algothrithms: 0(benchmark), 1(heuristic), 2(default). \n"
       "\t-q: [CUDA only] use separate stream for copy. \n"
+      "\t-z: Set denormal as zero. When tuning on this option reduces latency dramatically, a model may have denormals.\n"
```
Comment (tracysh) [inline onnxruntime/test/perftest/command_args_parser.cc:53]: tuning->turning?

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 3 above)
Comment (hanbitmyths) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:1101]: Add details here.

File: onnxruntime/core/common/denormal.cc [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+#if defined(_M_IX86) || defined(_M_X64) || defined(__i386__) || defined(__x86_64__)
+#define PLATFORM_X86
+#endif
+
+#if defined(PLATFORM_X86)
+#if defined(_MSC_VER)
+#include <intrin.h>
+#define DENORMAL_INTRINC
+// intrins headers at gcc 4.8 and older are not usable without compiler flags
+#elif defined(__GNUC__) && ((__GNUC__ >= 5) || ((__GNUC__ == 4) && (__GNUC_MINOR__ > 8)))
```
[the remark this answers is not in the document]
Comment (hanbitmyths) [reply onnxruntime/core/common/denormal.cc:13]: Add __clang__ not to be filtered by gcc 4.8 constraint. Clang works properly with intrinc.

File: onnxruntime/core/common/denormal.cc (hunk 5 above)
Comment (hanbitmyths) [reply onnxruntime/core/common/denormal.cc:61]: With firstprivate clause, all openmp threads are initialized with passing variable's copy constructor. It works even when loop count is less than maximum number of openmp threads. I put comments to explain a behavior.

File: onnxruntime/test/perftest/command_args_parser.cc (hunk 6 above)
Comment (hanbitmyths) [reply onnxruntime/test/perftest/command_args_parser.cc:53]: I also found the same typo from other places. Thanks for pointing out.

File: onnxruntime/core/common/denormal.cc [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+// intrins headers at gcc 4.8 and older are not usable without compiler flags.
+// clang claims gnuc 4.2, but it has a proper intrin header.
+#elif defined(__clang__) || (defined(__GNUC__) && ((__GNUC__ >= 5) || ((__GNUC__ == 4) && (__GNUC_MINOR__ > 8))))
+#include <cpuid.h>
+#include <x86intrin.h>
+#define DENORMAL_INTRINC
+#endif
+#endif
+
+#include <mutex>
```
Comment (tracysh) [inline onnxruntime/core/common/denormal.cc:21]: nit: is <mutex> still needed? I think you had it from std::call_once in an earlier iteration.

Comment (hanbitmyths) [reply onnxruntime/core/common/denormal.cc]: removed

File: include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h [hunk 9]
```diff
@@ -431,6 +432,7 @@ class ThreadPoolTempl : public onnxruntime::concurrency::ExtendedThreadPoolInter
 
     worker_data_.resize(num_threads_);
     for (int i = 0; i < num_threads_; i++) {
+      worker_data_[i].set_denormal_as_zero = thread_options.set_denormal_as_zero;
```
Comment (tracysh) [inline include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h:435]: (Sorry had this comment earlier, but I don't see what happened to it): Can set_denormal_as_zero just be a flag at the same level as allow_spinning_ and not need to be in the worker data?

Comment (hanbitmyths) [reply include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h]: Good point. Fixed.

**Review by tracysh — APPROVED** [review]

**Pull request merged**